### ISSUE TITLE
Any use of SSLHelper class is blocking (ie: HttpServer and HttpClient constructors)

### ISSUE THREAD

**Issue by ryanlamore** [vert-x3/issues#246]
Noticed a bunch of io.vertx.core.impl.BlockedThreadChecker exceptions happening on start up that I traced down to [SSLHelper](https://github.com/eclipse/vert.x/blob/master/src/main/java/io/vertx/core/net/impl/SSLHelper.java#L109) .  If you follow that code path down all the way, you'll find

> my.server.com    | Mar 16, 2017 5:20:25 PM io.vertx.core.impl.BlockedThreadChecker
> my.server.com    | WARNING: Thread Thread[vert.x-eventloop-thread-3,5,main] has been blocked for 9548 ms, time limit is 2000
> my.server.com    | io.vertx.core.VertxException: Thread blocked
> my.server.com    | 	at java.util.zip.Inflater.inflateBytes(Native Method)
> my.server.com    | 	at java.util.zip.Inflater.inflate(Inflater.java:259)
> my.server.com    | 	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:152)
> my.server.com    | 	at java.util.jar.JarVerifier$VerifierStream.read(JarVerifier.java:478)
> my.server.com    | 	at java.io.BufferedInputStream.read1(BufferedInputStream.java:284)
> my.server.com    | 	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
> my.server.com    | 	at javax.crypto.JarVerifier.verifySingleJar(JarVerifier.java:429)
> my.server.com    | 	at javax.crypto.JarVerifier.verifyJars(JarVerifier.java:322)
> my.server.com    | 	at javax.crypto.JarVerifier.verify(JarVerifier.java:250)
> my.server.com    | 	at javax.crypto.JceSecurity.verifyProviderJar(JceSecurity.java:160)
> my.server.com    | 	at javax.crypto.JceSecurity.getVerificationResult(JceSecurity.java:186)
> my.server.com    | 	at javax.crypto.JceSecurity.canUseProvider(JceSecurity.java:200)
> my.server.com    | 	at javax.crypto.Cipher.getInstance(Cipher.java:515)
> my.server.com    | 	at sun.security.ssl.JsseJce.getCipher(JsseJce.java:229)
> my.server.com    | 	at sun.security.ssl.CipherBox.<init>(CipherBox.java:179)
> my.server.com    | 	at sun.security.ssl.CipherBox.newCipherBox(CipherBox.java:263)
> my.server.com    | 	at sun.security.ssl.CipherSuite$BulkCipher.newCipher(CipherSuite.java:505)
> my.server.com    | 	at sun.security.ssl.CipherSuite$BulkCipher.isAvailable(CipherSuite.java:572)
> my.server.com    | 	at sun.security.ssl.CipherSuite$BulkCipher.isAvailable(CipherSuite.java:527)
> my.server.com    | 	at sun.security.ssl.CipherSuite.isAvailable(CipherSuite.java:194)
> my.server.com    | 	at sun.security.ssl.SSLContextImpl.getApplicableCipherSuiteList(SSLContextImpl.java:346)
> my.server.com    | 	at sun.security.ssl.SSLContextImpl.getDefaultCipherSuiteList(SSLContextImpl.java:297)
> my.server.com    | 	at sun.security.ssl.SSLEngineImpl.init(SSLEngineImpl.java:402)
> my.server.com    | 	at sun.security.ssl.SSLEngineImpl.<init>(SSLEngineImpl.java:349)
> my.server.com    | 	at sun.security.ssl.SSLContextImpl.engineCreateSSLEngine(SSLContextImpl.java:201)
> my.server.com    | 	at javax.net.ssl.SSLContext.createSSLEngine(SSLContext.java:329)
> my.server.com    | 	at io.vertx.core.net.impl.SSLHelper.<clinit>(SSLHelper.java:109)
> my.server.com    | 	at io.vertx.core.http.impl.HttpClientImpl.<init>(HttpClientImpl.java:78)

Anything using a BufferedInputStream to verify a jar is definitely a blocking call.  If you want to see the full stack trace, see the attachment
[threadsBlocked.txt](https://github.com/vert-x3/issues/files/861725/threadsBlocked.txt).

It was my understanding that creating an HttpClient or HttpServer is meant to be non-blocking and OK to be done within an event loop thread.  Is that assumption correct?  If so, than statically calling SSLContext.createSSLEngine() isn't correct.

**Comment by ryanlamore** [comment]
One very simple fix could be to move [the code in the static{} block initializing DEFAULT_JDK_CIPHER_SUITE](https://github.com/eclipse/vert.x/blob/master/src/main/java/io/vertx/core/net/impl/SSLHelper.java#L104) to [createContext](https://github.com/eclipse/vert.x/blob/master/src/main/java/io/vertx/core/net/impl/SSLHelper.java#L265); which is the ONLY place DEFAULT_JDK_CIPHER_SUITE is used and is already in a "run only once" context anyway.  Seems like a very simple change that would fix the problem.

Let me know if you agree with me and I'll create the pull request to fix it.

**Comment by IdanFridman** [comment]
anything new with the issue? we still experience this.  it doesnt occurs only on startups

**Comment by vietj** [comment]
can you show a stack trace @IdanFridman ? it looks like the fix suggested by @ryanlamore has been applied (load the default cipher suite in a static block), so it is not clear what happens with you.

**Comment by IdanFridman** [comment]
Julien,
Iam attaching stack trace. This wasnt happening during startup. It happens once in a while

Stacktrace:

2018-06-22 07:58:48,054 WARN  io.vertx.core.impl.BlockedThreadChecker - Thread Thread[vert.x-eventloop-thread-0,5,main] has been blocked for 1126 ms, time limit is 1000
io.vertx.core.VertxException: Thread blocked
	at java.lang.ClassLoader.findBootstrapClass(Native Method)
	at java.lang.ClassLoader.findBootstrapClassOrNull(ClassLoader.java:1015)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:413)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:411)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at sun.security.jca.ProviderConfig$2.run(ProviderConfig.java:215)
	at sun.security.jca.ProviderConfig$2.run(ProviderConfig.java:206)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.security.jca.ProviderConfig.doLoadProvider(ProviderConfig.java:206)
	at sun.security.jca.ProviderConfig.getProvider(ProviderConfig.java:187)
	at sun.security.jca.ProviderList.getProvider(ProviderList.java:233)
	at sun.security.jca.ProviderList.getIndex(ProviderList.java:263)
	at sun.security.jca.ProviderList.getProviderConfig(ProviderList.java:247)
	at sun.security.jca.ProviderList.getProvider(ProviderList.java:253)
	at java.security.Security.getProvider(Security.java:468)
	at sun.security.ssl.SignatureAndHashAlgorithm.<clinit>(SignatureAndHashAlgorithm.java:415)
	at sun.security.ssl.SSLSessionImpl.<init>(SSLSessionImpl.java:179)
	at sun.security.ssl.SSLSessionImpl.<init>(SSLSessionImpl.java:147)
	at sun.security.ssl.SSLSessionImpl.<clinit>(SSLSessionImpl.java:79)
	at sun.security.ssl.SSLEngineImpl.init(SSLEngineImpl.java:369)
	at sun.security.ssl.SSLEngineImpl.<init>(SSLEngineImpl.java:349)
	at sun.security.ssl.SSLContextImpl$AbstractTLSContext.createSSLEngineImpl(SSLContextImpl.java:453)
	at sun.security.ssl.SSLContextImpl.engineCreateSSLEngine(SSLContextImpl.java:194)
	at javax.net.ssl.SSLContext.createSSLEngine(SSLContext.java:329)
	at io.vertx.core.net.impl.SSLHelper.<clinit>(SSLHelper.java:106)
	at io.vertx.core.net.impl.NetServerImpl.<init>(NetServerImpl.java:87)
	at io.vertx.core.impl.VertxImpl.createNetServer(VertxImpl.java:253)
	at io.vertx.core.eventbus.impl.clustered.ClusteredEventBus.lambda$start$1(ClusteredEventBus.java:138)
	at io.vertx.core.eventbus.impl.clustered.ClusteredEventBus$$Lambda$36/1705613512.handle(Unknown Source)
	at io.vertx.core.impl.FutureImpl.setHandler(FutureImpl.java:76)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:289)
	at io.vertx.core.impl.ContextImpl$$Lambda$17/435354399.handle(Unknown Source)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:339)
	at io.vertx.core.impl.ContextImpl$$Lambda$18/1063752855.run(Unknown Source)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

**Comment by ryanlamore** [comment]
@vietj : I don't think you understood me correctly.  I'm saying it should **NOT** be created in a static block.  As it is using a BufferedInputStream.  The problem still exists.

**Comment by vietj** [comment]
@ryanlamore that's something that is beyond control of Vert.x (SSL engine initialization), this happens during the startup phase of your application (so if you use bare Netty, you will have the same issue).

if you have concerns, you can wrap the initialization in an executeBlocking call.

**Comment by ryanlamore** [comment]
@vietj : I totally disagree with you.  Isn't [io.vertx.core.net.impl.SSLHelper](https://github.com/eclipse/vert.x/blob/master/src/main/java/io/vertx/core/net/impl/SSLHelper.java#L106) owned by vertx?  All I'm saying is when you statically call this method (SSLContext.getInstance("TLS").context.createSSLEngine()), you end up calling a blocking API.  The only way to get around this is to wrap [VertxImpl.createHttpServer()](https://github.com/eclipse/vert.x/blob/master/src/main/java/io/vertx/core/impl/VertxImpl.java#L286) in an executeBlocking.  But if that's the case, shouldn't **createHttpServer** be renamed to **createHttpServerBlocking**() to make it clear that this is a blocking call?

**Comment by vietj** [comment]
@ryanlamore have you experienced actual issues due to this ?

**Comment by ryanlamore** [comment]
@vietj : Besides error stack traces in our logs, no.  But we take our logs very seriously and it's unfortunate that we just have to **live** with these stack traces.  I'm happy to make the fix and send the PR.  If I do so and you like the change, can we get it merged into the codebase for some future release?

**Comment by vietj** [comment]
@ryanlamore it is complicated to find the right solution to this problem, what do you have in mind ?
